docs/cudf/source/developer_guide/cudf_pandas.md

```diff
@@ -0,0 +1,119 @@
+
+##### Attributes and Callable Proxy Types:
+Final proxy types are typically classes or modules, both of which have attributes.
+Classes also have methods.
+These attributes and methods must be wrapped as well to support the fast-slow proxy scheme.
+
+#### Creating New Proxy Types:
+`_FinalProxy` and `_IntermediateProxy` types are created using the functions `make_final_proxy_type` and `make_intermediate_proxy` type, respectively. Creating a new final type looks like this.
+
+```python
+DataFrame = make_final_proxy_type(
+    "DataFrame",
+    cudf.DataFrame,
+    pd.DataFrame,
+    fast_to_slow=lambda fast: fast.to_pandas(),
+    slow_to_fast=cudf.from_pandas,
+)
+```
+
+### The Fallback Mechanism
+The `_fast_slow_function_call` is the all-important function that we use to call operations the fast way (using cuDF) and if that fails, the slow way (using Pandas).
+The is also known as the fallback mechanism.
+The function looks like this:
+```python
+def _fast_slow_function_call(func: Callable, *args, **kwargs):
+    try:
+        ...
+        fast_args, fast_kwargs = _fast_arg(args), _fast_arg(kwargs)
+        result = func(*fast_args, **fast_kwargs)
+        ...
+    except Exception:
+        ...
+        slow_args, slow_kwargs = _slow_arg(args), _slow_arg(kwargs)
+        result = func(*slow_args, **slow_kwargs)
+        ...
+    return _maybe_wrap_result(result, func, *args, **kwargs), fast
+```
+As you can see the function attempts to call `func` the fast way using cuDF and if any `Exception` occurs, it calls the function using Pandas.
+In essence, this `try-except` is what allows `cudf.pandas` to support 100% of the Pandas API.
+
+At the end, the function wraps the result from either path in a fast-slow proxy object, if necessary.
+
+#### Converting Proxy Objects
+Note that before the `func` is called, the proxy object and its attributes need to be converted to either their cuDF or Pandas implementations.
+This conversion is handled in the function `_transform_arg` which both `_fast_arg` and `_slow_arg` call.
+
+`_transform_arg` is a recuirsive function that will call itself depending on the type or argument passed to it (eg. `_transform_arg` is called for each element in a list of arguments).
+
+### Using Metaclasses
+`cudf.pandas` uses a [metaclass](https://docs.python.org/3/glossary.html#term-metaclass) called (`_FastSlowProxyMeta`) to dynamically find class attributes and classmethods of fast-slow proxy types.
+For example, in the snippet below, the `xpd.Series` type is an instance of `_FastSlowProxyMeta`.
+Therefore we can access the property `_fsproxy_fast` defined in the metaclass.
+```python
+import cudf.pandas
+cudf.pandas.install()
+import pandas as xpd
+
+print(xpd.Series._fsproxy_fast) # output is cudf.core.series.Series
+```
+
+## debugging `cudf.pandas`
+Several environment variables are available for debugging purposes.
```

Comment:
*question*: What are the other ones?

Reply:
The next one (or possibly ones) I'm adding is for debugging different types of fallback (eg. OOM errors, NotImplementedErrors). I'm working on it #15910.

Reply:
Ah thanks.